### DISCUSSION TITLE
on_bar不能回调的原因

### DISCUSSION THREAD

**Discussion by zhuzhenping** [wondertrader/wondertrader#109] [Q&A] [answered]
根据Wesley回答网友问题整理.

**Comment by zhuzhenping** [comment]
datakit实际上只是一个python封装

datakit是利用WtDtPorter对WtDtCore的封装

而QuoteFacotry就是对WtDtCore的C++封装

**Comment by zhuzhenping** [comment]
QuoteFacotry本质上和datakit没有区别

无非就是调用的入口不同

**Reply by wondertrader** [reply]
是的

**Comment by zhuzhenping** [comment]
问：WtCtaRtTicker::on_tick()方法里没找到创建和更新bar的代码，请问它们在什么地方？

答：WtDataReader，onMinuteEnd

问：就是在这个方法里没找着

答：这是Reader啊，你看WtDataWriter，updateMinuteData

问：这个方法我看见了，这哪调用呢？

答：数据落地和交易是两个进程，这样即使盘中干预交易进程，也不会影响数据落地，实际上，你需要从WtDtCore里去看

问：我找一下,就是说，运行WtDtRunner进行数据落地？

答：对，QuoteFactory

问：数据落地的main()在哪儿？

答：QuoteFactory

问：这个大概什么时候有更新啊,这个方法在哪调用？WTSBarStruct* WTSDataFactory::updateKlineData(WTSKlineData* klineData, WTSTickData* tick, WTSSessionInfo* sInfo)

答：on_tick吧，WtDtCore/DataManager/WtDataWriter

问：我的版本不对吗？我的WtDataWriter是个单独目录

答：是单独项目啊，我是说调用路径

问：WtDataWriter没调用updateKlineData方法，没搜索到调用

答：哦，这个方法是不调用的，这个方法我没用到，这是我以前写的，拼K线的方法，算是白放进去的。

问：那怎么用tick更新bar呢？

答：你看WtDataWriter，数据落地都在这里

**Comment by zhuzhenping** [comment]
问：
是不是如果不启动datakit，就无法接收到行情啊，那cpp版本的好像没有这个东西

答：
cpp版叫做QuoteFactory

问：
需要启动吗，我昨天做仿真好像没有这个

答:
要么启动datakit,要么启动quotefactory,二选一

问：
cpp版里是自己启动了吗，我好像没有单独启动这个

答：
要单独启动。QuoteFactory和datakit只是入口不同，内部一样

问：
请问下这config.ini在哪里啊。。。

答：
我可以把QuoteFactory的配置文件名改得和datakit一样，这样就可以互相替换

问：
我昨天仿真的时候没启动这个，但是能够正常运行呀

答：
可能你启动了datakit吧，二选一的，datakit是python入口的，qf是自己就是入口
parser都是可以替换的
parser只是一个组件，随便换都行
但是！！
如果要用K线数据，就依赖datakit
因为datakit/qf才是处理数据落地的服务

parser只传回tick行情，datakit负责聚合成K线？

对

datakit会开一个udp广播
所以parserudp实际上是从datakit收数据的
但是如果你对落地的数据和你接收的数据的同步性要求不高
那么你甚至可以runner直连ctp
K线数据从datakit去拿就好了

**Comment by zhuzhenping** [comment, marked as answer]
一、看一下on_init函数是否订阅K线(stra_get_bars)。

二、没有启动datakit。直连ctp，没有数据落地，不会生成K线，需要启动QuoteFactory。要启动datakit，然后再启动quotefactory，配置WtRunner的parser为parserudp。on_bar一定是K线闭合了才回调。

UDP不广播分钟线，但是会触发WtDataReader去读分钟线，不推送信息，主要是时间戳，靠时间戳触发读取的逻辑，
定时吗？
不是。

那这时间戳谁来触发？
ParserUDP接收datakit广播的tick数据，tick带有交易所的时间戳。

然后WtDataReader会在每分钟结束的时候，更新最新的K线，如果K线闭合，就会触发onbar。

读的还是datakit存储的共享内存K线是吧？
对的mmap，历史K线直接缓存到内存中，实时K线就是mmap。

ParserUDP，一般交易用

datakit不用

ParserUDP是接收从datakit广播出来的行情

UDPCaster 呢？

广播用的啊

高频交易用udp么？

高频最好还是不用吧

udp的好处，是一个datakit可以给多个交易进程用

**Reply by wondertrader** [reply]
你这问题整理得不错啊

**Reply by zhuzhenping** [reply]
> 你这问题整理得不错啊

都是您以前亲自写的答复，正好这几天我也碰到了这个问题，好好的学习一下！

**Reply by ywang862** [reply]
为什么要用udp来做多进程通信，不用shm？高频系统内部延迟确定只有几微？udp走一走就不止了吧

**Comment by zhuzhenping** [comment]
当1分钟和5分钟K线在on_bar中正常生成，多分钟K线不能正常生成时，看看sessions.json是否修改过，用正常的sessions.json文件替换。
